In [ ]:
pip install mlxtend

In [2]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
import matplotlib.pyplot as plt
import networkx as nx
df = pd.read_csv('Customer Preference Survey (Responses).csv')

In [3]:
df

Timestamp playstation5 เครื่องทำขนมปัง Ergonomic Wrist Rest  \
0   10/17/2021 17:29:59          NaN             NaN                  NaN   
1   10/17/2021 17:43:38   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
2   10/17/2021 17:43:41   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
3   10/17/2021 17:43:46   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
4   10/17/2021 17:43:49   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
5   10/17/2021 17:43:55   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
6   10/17/2021 17:43:56   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
7   10/17/2021 17:43:57   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
8   10/17/2021 17:44:06   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
9   10/17/2021 17:44:07   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
10  10/17/2021 17:44:07   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
11  10/17/2021 17:44:07   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
12  10/17/2021 17:44:08   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
13  10/17/2021 17:44:12   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
14  10/17/2021 17:44:15   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
15  10/17/2021 17:44:18   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
16  10/17/2021 17:44:18   ไม่เคยซื้อ      ไม่เคยซื้อ              เคยซื้อ   
17  10/17/2021 17:44:20   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
18  10/17/2021 17:44:22   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
19  10/17/2021 17:44:26   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
20  10/17/2021 17:44:27   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
21  10/17/2021 17:44:27   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
22  10/17/2021 17:44:27   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
23  10/17/2021 17:44:29   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
24  10/17/2021 17:44:30      เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
25  10/17/2021 17:44:33   ไม่เคยซื้อ      ไม่เคยซื้อ              เคยซื้อ   
26  10/17/2021 17:44:35   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
27  10/17/2021 17:44:36   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
28  10/17/2021 17:44:37   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
29  10/17/2021 17:44:46   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   
30  10/17/2021 17:44:47   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
31  10/17/2021 17:44:49   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   
32  10/17/2021 17:44:55   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
33  10/17/2021 17:44:58   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
34  10/17/2021 17:45:00   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
35  10/17/2021 17:45:05   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
36  10/17/2021 17:45:05   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
37  10/17/2021 17:45:13   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
38  10/17/2021 17:45:13   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
39  10/17/2021 17:45:27   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
40  10/17/2021 17:45:39   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
41  10/17/2021 17:45:50   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
42  10/17/2021 17:46:06   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   
43  10/17/2021 17:46:07   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
44  10/17/2021 17:46:17   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
45  10/17/2021 17:46:29   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
46  10/17/2021 17:46:31   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   

   เครื่องอบผ้า เครื่องชงกาแฟแคปซูล เก้าอี้ LA-Z-Boy  \
0           NaN                 NaN              NaN   
1    ไม่เคยซื้อ          ไม่เคยซื้อ       ไม่เคยซื้อ   
2    ไม่เคยซื้อ          ไม่เคยซื้อ       ไม่เคยซื้อ   
3       เคยซื้อ             เคยซื้อ       ไม่เคยซื้อ   
4    ไม่เคยซื้อ          ไม่เคยซื้อ  

In [4]:
#Drop first row
df = df.drop(0)
df

Timestamp playstation5 เครื่องทำขนมปัง Ergonomic Wrist Rest  \
1   10/17/2021 17:43:38   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
2   10/17/2021 17:43:41   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
3   10/17/2021 17:43:46   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
4   10/17/2021 17:43:49   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
5   10/17/2021 17:43:55   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
6   10/17/2021 17:43:56   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
7   10/17/2021 17:43:57   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
8   10/17/2021 17:44:06   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
9   10/17/2021 17:44:07   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
10  10/17/2021 17:44:07   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
11  10/17/2021 17:44:07   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
12  10/17/2021 17:44:08   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
13  10/17/2021 17:44:12   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
14  10/17/2021 17:44:15   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
15  10/17/2021 17:44:18   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
16  10/17/2021 17:44:18   ไม่เคยซื้อ      ไม่เคยซื้อ              เคยซื้อ   
17  10/17/2021 17:44:20   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
18  10/17/2021 17:44:22   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
19  10/17/2021 17:44:26   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
20  10/17/2021 17:44:27   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
21  10/17/2021 17:44:27   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
22  10/17/2021 17:44:27   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
23  10/17/2021 17:44:29   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
24  10/17/2021 17:44:30      เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
25  10/17/2021 17:44:33   ไม่เคยซื้อ      ไม่เคยซื้อ              เคยซื้อ   
26  10/17/2021 17:44:35   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
27  10/17/2021 17:44:36   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
28  10/17/2021 17:44:37   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
29  10/17/2021 17:44:46   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   
30  10/17/2021 17:44:47   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
31  10/17/2021 17:44:49   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   
32  10/17/2021 17:44:55   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
33  10/17/2021 17:44:58   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
34  10/17/2021 17:45:00   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
35  10/17/2021 17:45:05   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
36  10/17/2021 17:45:05   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
37  10/17/2021 17:45:13   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
38  10/17/2021 17:45:13   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
39  10/17/2021 17:45:27   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
40  10/17/2021 17:45:39   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
41  10/17/2021 17:45:50   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
42  10/17/2021 17:46:06   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   
43  10/17/2021 17:46:07   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
44  10/17/2021 17:46:17   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
45  10/17/2021 17:46:29   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   
46  10/17/2021 17:46:31   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   

   เครื่องอบผ้า เครื่องชงกาแฟแคปซูล เก้าอี้ LA-Z-Boy  \
1    ไม่เคยซื้อ          ไม่เคยซื้อ       ไม่เคยซื้อ   
2    ไม่เคยซื้อ          ไม่เคยซื้อ       ไม่เคยซื้อ   
3       เคยซื้อ             เคยซื้อ       ไม่เคยซื้อ   
4    ไม่เคยซื้อ          ไม่เคยซื้อ       ไม่เคยซื้อ   
5       เคยซื้อ          ไม่เคยซื้อ          เคยซื้อ   
6    ไม่เคยซื้อ          ไม่เคยซื้อ       ไม่เคยซื้อ   
7 

In [5]:
#drop duplicated PS5 column
df_clean = df.drop(columns=['PS5', 'Timestamp'])
df_clean

playstation5 เครื่องทำขนมปัง Ergonomic Wrist Rest เครื่องอบผ้า  \
1    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
2    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
3    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
4    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
5    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
6    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
7    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
8    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
9    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
10   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
11   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
12   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
13   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
14   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
15   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
16   ไม่เคยซื้อ      ไม่เคยซื้อ              เคยซื้อ   ไม่เคยซื้อ   
17   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
18   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
19   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
20   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
21   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
22   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
23   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
24      เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
25   ไม่เคยซื้อ      ไม่เคยซื้อ              เคยซื้อ   ไม่เคยซื้อ   
26   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
27   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
28   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
29   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
30   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
31   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
32   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
33   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
34   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
35   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
36   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
37   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
38   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
39   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
40   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
41   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
42   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
43   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
44   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
45   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
46   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   

   เครื่องชงกาแฟแคปซูล เก้าอี้ LA-Z-Boy เครื่องให้อาหารสัตว์อัตโนมัติ  \
1           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
2           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
3              เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
4           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
5           ไม่เคยซื้อ          เคยซื้อ                        ไม่เคย   
6           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
7           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
8              เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
9           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
10          ไม่เคยซื้อ       

In [6]:
#fillna
df_clean = df_clean.fillna(0)
df_clean

playstation5 เครื่องทำขนมปัง Ergonomic Wrist Rest เครื่องอบผ้า  \
1    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
2    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
3    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
4    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
5    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
6    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
7    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
8    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
9    ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
10   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
11   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
12   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
13   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
14   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
15   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
16   ไม่เคยซื้อ      ไม่เคยซื้อ              เคยซื้อ   ไม่เคยซื้อ   
17   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
18   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
19   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
20   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
21   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
22   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
23   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
24      เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
25   ไม่เคยซื้อ      ไม่เคยซื้อ              เคยซื้อ   ไม่เคยซื้อ   
26   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
27   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
28   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
29   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
30   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
31   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
32   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ      เคยซื้อ   
33   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
34   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
35   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
36   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
37   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
38   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
39   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
40   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
41   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
42   ไม่เคยซื้อ         เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
43   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
44   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
45   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   
46   ไม่เคยซื้อ      ไม่เคยซื้อ           ไม่เคยซื้อ   ไม่เคยซื้อ   

   เครื่องชงกาแฟแคปซูล เก้าอี้ LA-Z-Boy เครื่องให้อาหารสัตว์อัตโนมัติ  \
1           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
2           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
3              เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
4           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
5           ไม่เคยซื้อ          เคยซื้อ                        ไม่เคย   
6           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
7           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
8              เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
9           ไม่เคยซื้อ       ไม่เคยซื้อ                        ไม่เคย   
10          ไม่เคยซื้อ       

In [7]:
#แปลงเคยซื้อ ไม่เคยซื้อเป็น 0,1
df_clean=df_clean.replace('ไม่เคยซื้อ',0)
df_clean=df_clean.replace('ไม่เคย',0)
df_clean=df_clean.replace('ไม่',0)
df_clean=df_clean.replace('ไม่เคยซือ',0)
df_clean=df_clean.replace('เคยซื้อ',1)
df_clean=df_clean.replace('เคย',1)
#df_clean.iloc[1].unique()
df_clean

playstation5  เครื่องทำขนมปัง  Ergonomic Wrist Rest  เครื่องอบผ้า  \
1              0                0                     0             0   
2              0                0                     0             0   
3              0                0                     0             1   
4              0                0                     0             0   
5              0                0                     0             1   
6              0                0                     0             0   
7              0                0                     0             0   
8              0                0                     0             1   
9              0                0                     0             0   
10             0                0                     0             0   
11             0                0                     0             0   
12             0                0                     0             0   
13             0                0                     0             0   
14             0                0                     0             1   
15             0                0                     0             1   
16             0                0                     1             0   
17             0                0                     0             0   
18             0                0                     0             0   
19             0                0                     0             0   
20             0                0                     0             0   
21             0                0                     0             0   
22             0                0                     0             0   
23             0                0                     0             0   
24             1                0                     0             0   
25             0                0                     1             0   
26             0                0                     0             0   
27             0                0                     0             0   
28             0                0                     0             0   
29             0                1                     0             0   
30             0                0                     0             0   
31             0                1                     0             0   
32             0                0                     0             1   
33             0                0                     0             0   
34             0                0                     0             0   
35             0                0                     0             0   
36             0                0                     0             0   
37             0                0                     0             0   
38             0                0                     0             0   
39             0                0                     0             0   
40             0                0                     0             0   
41             0                0                     0             0   
42             0                1                     0             0   
43             0                0                     0             0   
44             0                0                     0             0   
45             0                0                     0             0   
46             0                0                     0             0   

    เครื่องชงกาแฟแคปซูล  เก้าอี้ LA-Z-Boy  เครื่องให้อาหารสัตว์อัตโนมัติ  \
1                     0                 0                              0   
2                     0                 0                              0   
3                     1                 0                              0   
4                     0                 0                              0   
5                     0                 1                              0   
6                     0                 0                              0   
7                     0                 

In [8]:
basket_encode_final_df = df_clean[(df_clean>0).sum(axis=1) >= 2]
basket_encode_final_df

playstation5  เครื่องทำขนมปัง  Ergonomic Wrist Rest  เครื่องอบผ้า  \
1              0                0                     0             0   
2              0                0                     0             0   
3              0                0                     0             1   
4              0                0                     0             0   
5              0                0                     0             1   
6              0                0                     0             0   
7              0                0                     0             0   
8              0                0                     0             1   
9              0                0                     0             0   
10             0                0                     0             0   
11             0                0                     0             0   
12             0                0                     0             0   
13             0                0                     0             0   
14             0                0                     0             1   
15             0                0                     0             1   
16             0                0                     1             0   
17             0                0                     0             0   
18             0                0                     0             0   
19             0                0                     0             0   
20             0                0                     0             0   
21             0                0                     0             0   
22             0                0                     0             0   
23             0                0                     0             0   
24             1                0                     0             0   
25             0                0                     1             0   
26             0                0                     0             0   
27             0                0                     0             0   
28             0                0                     0             0   
29             0                1                     0             0   
30             0                0                     0             0   
31             0                1                     0             0   
32             0                0                     0             1   
33             0                0                     0             0   
34             0                0                     0             0   
35             0                0                     0             0   
36             0                0                     0             0   
37             0                0                     0             0   
38             0                0                     0             0   
39             0                0                     0             0   
40             0                0                     0             0   
41             0                0                     0             0   
42             0                1                     0             0   
43             0                0                     0             0   
44             0                0                     0             0   
45             0                0                     0             0   
46             0                0                     0             0   

    เครื่องชงกาแฟแคปซูล  เก้าอี้ LA-Z-Boy  เครื่องให้อาหารสัตว์อัตโนมัติ  \
1                     0                 0                              0   
2                     0                 0                              0   
3                     1                 0                              0   
4                     0                 0                              0   
5                     0                 1                              0   
6                     0                 0                              0   
7                     0                 

In [9]:
freq_items_set = apriori(basket_encode_final_df,
                        min_support=0.1,
                        use_colnames=True).sort_values('support', ascending=False).reset_index(drop=True)

freq_items_set

support                                           itemsets
0      0.847826                                   (Salmon Sashimi)
1      0.826087                                             (ยาดม)
2      0.804348                                (External Harddisk)
3      0.739130                                (Bluetooth Speaker)
4      0.739130                             (Salmon Sashimi, ยาดม)
...         ...                                                ...
10686  0.108696  (เครื่องชงกาแฟแคปซูล, Robot ดูดฝุ่น, ขนมจีนน้ำ...
10687  0.108696  (เครื่องชงกาแฟแคปซูล, Robot ดูดฝุ่น, ขนมจีนน้ำ...
10688  0.108696  (เครื่องชงกาแฟแคปซูล, Robot ดูดฝุ่น, ขนมจีนน้ำ...
10689  0.108696  (เครื่องชงกาแฟแคปซูล, Robot ดูดฝุ่น, แก้วเก็บค...
10690  0.108696  (Robot ดูดฝุ่น, ขนมจีนน้ำยาปู, หม้อทอดไร้น้ําม...

[10691 rows x 2 columns]

In [13]:
rules = association_rules(freq_items_set, metric="lift", min_threshold=1).sort_values('lift', ascending=False).reset_index(drop=True)

rules['consequents_lenght'] = rules['consequents'].apply(lambda x : len(x))
rules['antecedents_lenght'] = rules['antecedents'].apply(lambda x : len(x))

rules

antecedents  \
0       (Robot ดูดฝุ่น, หม้อทอดไร้น้ํามัน, แก้วเก็บควา...   
1       (เครื่องซักผ้า, Salmon Sashimi, ลู่วิ่งออกกำลั...   
2                                               (ทรายแมว)   
3       (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, แก้วเก็บควา...   
4       (ขนมจีนน้ำยาปู, airpods , พลาสเตอร์บรรเทาปวด ต...   
...                                                   ...   
781579          (Salmon Sashimi, External Harddisk, ยาดม)   
781580                                   (หนังสือ python)   
781581                                   (หนังสือ python)   
781582  (แก้วเก็บความเย็น, พลาสเตอร์บรรเทาปวด ตราเสือ,...   
781583                          (Deskmat, หนังสือ python)   

                                              consequents  antecedent support  \
0       (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, เครื่องซักผ...            0.130435   
1       (ขนมจีนน้ำยาปู, airpods , พลาสเตอร์บรรเทาปวด ต...            0.108696   
2                                              (น้ำพุแมว)            0.130435   
3       (Robot ดูดฝุ่น, Mechanical keyboard, ยาดม, หม้...            0.108696   
4               (เครื่องซักผ้า, ยาดม, ลู่วิ่งออกกำลังกาย)            0.130435   
...                                                   ...                 ...   
781579                              (Mechanical keyboard)            0.608696   
781580  (ขนมจีนน้ำยาปู, Salmon Sashimi, Mechanical key...            0.500000   
781581  (ขนมจีนน้ำยาปู, ยาดม, เครื่องซักผ้า, หม้อทอดไร...            0.500000   
781582                              (Mechanical keyboard)            0.260870   
781583  (Bluetooth Speaker, Salmon Sashimi, External H...            0.217391   

        consequent support   support  confidence      lift  leverage  \
0                 0.108696  0.108696    0.833333  7.666667  0.094518   
1                 0.130435  0.108696    1.000000  7.666667  0.094518   
2                 0.108696  0.108696    0.833333  7.666667  0.094518   
3                 0.130435  0.108696    1.000000  7.666667  0.094518   
4                 0.108696  0.108696    0.833333  7.666667  0.094518   
...                    ...       ...         ...       ...       ...   
781579            0.500000  0.304348    0.500000  1.000000  0.000000   
781580            0.217391  0.108696    0.217391  1.000000  0.000000   
781581            0.217391  0.108696    0.217391  1.000000  0.000000   
781582            0.500000  0.130435    0.500000  1.000000  0.000000   
781583            0.500000  0.108696    0.500000  1.000000  0.000000   

        conviction  consequents_lenght  antecedents_lenght  
0         5.347826                   5                   5  
1              inf                   4                   3  
2         5.347826                   1                   1  
3              inf                   4                   5  
4         5.347826                   3                   4  
...            ...                 ...                 ...  
781579    1.000000                   1                   3  
781580    1.000000                   4                   1  
781581    1.000000                   4                   1  
781582    1.000000                   1                   3  
781583    1.000000                   4                   2  

[781584 rows x 11 columns]

In [32]:
rules[(rules['antecedents_lenght'] <= 1) &
      (rules['consequents_lenght'] <= 1) &
      (rules['lift'] >= 2) &
      (rules['support'] >= 0.1) &
      (rules['confidence'] >= 0.5) ]


antecedents            consequents  \
2                          (ทรายแมว)             (น้ำพุแมว)   
9                         (น้ำพุแมว)              (ทรายแมว)   
62513             (กระติกน้ำ 2 ลิตร)         (เครื่องอบผ้า)   
62515                 (เครื่องอบผ้า)     (กระติกน้ำ 2 ลิตร)   
314614  (Logitech Mx Master 3 Mouse)  (เครื่องชงกาแฟแคปซูล)   
445767             (Nintendo switch)       (หนังสือ python)   
448504                (gaming chair)  (Mechanical keyboard)   
448774            (ไฟส่องหน้าไลฟ์สด)  (Mechanical keyboard)   

        antecedent support  consequent support   support  confidence  \
2                 0.130435            0.108696  0.108696    0.833333   
9                 0.108696            0.130435  0.108696    1.000000   
62513             0.217391            0.130435  0.108696    0.500000   
62515             0.130435            0.217391  0.108696    0.833333   
314614            0.130435            0.347826  0.108696    0.833333   
445767            0.130435            0.500000  0.130435    1.000000   
448504            0.108696            0.500000  0.108696    1.000000   
448774            0.130435            0.500000  0.130435    1.000000   

            lift  leverage  conviction  consequents_lenght  antecedents_lenght  
2       7.666667  0.094518    5.347826                   1                   1  
9       7.666667  0.094518         inf                   1                   1  
62513   3.833333  0.080340    1.739130                   1                   1  
62515   3.833333  0.080340    4.695652                   1                   1  
314614  2.395833  0.063327    3.913043                   1                   1  
445767  2.000000  0.065217         inf                   1                   1  
448504  2.000000  0.054348         inf                   1                   1  
448774  2.000000  0.065217         inf                   1                   1

In [33]:
rules.head(60)

antecedents  \
0   (Robot ดูดฝุ่น, หม้อทอดไร้น้ํามัน, แก้วเก็บควา...   
1   (เครื่องซักผ้า, Salmon Sashimi, ลู่วิ่งออกกำลั...   
2                                           (ทรายแมว)   
3   (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, แก้วเก็บควา...   
4   (ขนมจีนน้ำยาปู, airpods , พลาสเตอร์บรรเทาปวด ต...   
5                 (เครื่องซักผ้า, ลู่วิ่งออกกำลังกาย)   
6   (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, แก้วเก็บควา...   
7   (เครื่องซักผ้า, External Harddisk, ลู่วิ่งออกก...   
8   (ขนมจีนน้ำยาปู, หม้อทอดไร้น้ํามัน, airpods , พ...   
9                                          (น้ำพุแมว)   
10  (ขนมจีนน้ำยาปู, หม้อทอดไร้น้ํามัน, airpods , S...   
11                (เครื่องซักผ้า, ลู่วิ่งออกกำลังกาย)   
12  (ขนมจีนน้ำยาปู, airpods , พลาสเตอร์บรรเทาปวด ต...   
13  (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, เครื่องซักผ...   
14  (ขนมจีนน้ำยาปู, airpods , พลาสเตอร์บรรเทาปวด ต...   
15  (Bluetooth Speaker, พลาสเตอร์บรรเทาปวด ตราเสือ...   
16                (เครื่องซักผ้า, ลู่วิ่งออกกำลังกาย)   
17  (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, เครื่องซักผ...   
18  (หม้อทอดไร้น้ํามัน, Bluetooth Speaker, Mechani...   
19  (Bluetooth Speaker, พลาสเตอร์บรรเทาปวด ตราเสือ...   
20  (เครื่องซักผ้า, Salmon Sashimi, ยาดม, ลู่วิ่งอ...   
21  (เครื่องซักผ้า, Salmon Sashimi, External Hardd...   
22                        (แก้วเก็บความเย็น, ทรายแมว)   
23  (ขนมจีนน้ำยาปู, หม้อทอดไร้น้ํามัน, airpods , ย...   
24  (ลู่วิ่งออกกำลังกาย, เครื่องซักผ้า, Salmon Sas...   
25  (ขนมจีนน้ำยาปู, airpods , พลาสเตอร์บรรเทาปวด ต...   
26                       (แก้วเก็บความเย็น, น้ำพุแมว)   
27                                          (ทรายแมว)   
28  (ขนมจีนน้ำยาปู, หม้อทอดไร้น้ํามัน, airpods , S...   
29  (หม้อทอดไร้น้ํามัน, Bluetooth Speaker, Mechani...   
30  (ขนมจีนน้ำยาปู, airpods , Salmon Sashimi, ยาดม...   
31  (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, แก้วเก็บควา...   
32  (ขนมจีนน้ำยาปู, หม้อทอดไร้น้ํามัน, airpods , S...   
33  (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, แก้วเก็บควา...   
34  (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, แก้วเก็บควา...   
35          (เครื่องซักผ้า, ยาดม, ลู่วิ่งออกกำลังกาย)   
36  (เครื่องซักผ้า, Salmon Sashimi, airpods , หนัง...   
37  (ขนมจีนน้ำยาปู, airpods , พลาสเตอร์บรรเทาปวด ต...   
38  (ขนมจีนน้ำยาปู, airpods , พลาสเตอร์บรรเทาปวด ต...   
39  (Robot ดูดฝุ่น, Mechanical keyboard, ยาดม, หม้...   
40  (Robot ดูดฝุ่น, หม้อทอดไร้น้ํามัน, ยาดม, Mecha...   
41  (ขนมจีนน้ำยาปู, หม้อทอดไร้น้ํามัน, airpods , ย...   
42  (เครื่องซักผ้า, External Harddisk, ลู่วิ่งออกก...   
43  (Robot ดูดฝุ่น, Mechanical keyboard, ยาดม, หม้...   
44  (Robot ดูดฝุ่น, หม้อทอดไร้น้ํามัน, ยาดม, Mecha...   
45            (Bluetooth Speaker, ลู่วิ่งออกกำลังกาย)   
46  (ขนมจีนน้ำยาปู, airpods , พลาสเตอร์บรรเทาปวด ต...   
47  (airpods , เครื่องซักผ้า, Salmon Sashimi, ยาดม...   
48          (เครื่องซักผ้า, ยาดม, ลู่วิ่งออกกำลังกาย)   
49  (airpods , เครื่องซักผ้า, Salmon Sashimi, Exte...   
50  (Robot ดูดฝุ่น, หม้อทอดไร้น้ํามัน, เครื่องซักผ...   
51  (airpods , เครื่องซักผ้า, Salmon Sashimi, Exte...   
52  (เครื่องซักผ้า, Salmon Sashimi, External Hardd...   
53  (เครื่องซักผ้า, ยาดม, External Harddisk, ลู่วิ...   
54  (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, Salmon Sash...   
55  (หม้อทอดไร้น้ํามัน, Bluetooth Speaker, Mechani...   
56  (Bluetooth Speaker, พลาสเตอร์บรรเทาปวด ตราเสือ...   
57  (Robot ดูดฝุ่น, หม้อทอดไร้น้ํามัน, แก้วเก็บควา...   
58  (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, แก้วเก็บควา...   
59  (Bluetooth Speaker, Salmon Sashimi, ลู่วิ่งออก...   

                                          consequents  antecedent support  \
0   (ขนมจีนน้ำยาปู, ไฟแต่งห้องมินิมอล, เครื่องซักผ...            0.130435   
1   (ขนมจีนน้ำยาปู, airpods , พลาสเตอร์บรรเทาปวด ต...            0.108696   
2                                          (น้ำพุแมว)            0.130435   
3   (Robot ดูดฝุ่น, Mechanical keyboard, ยาดม, หม้...            0.108696   
4           (เครื่องซักผ้า, ยาดม, ลู่วิ่งออกกำลังกาย)            0.130435   
5   (ขนมจีนน้ำยาปู, หม้อทอดไร้น้ํามัน, airpods , S...            0.108696   
6   (Robot ดูดฝุ่น, Mecha

# Top 10 Product Recommendation

1. หมวดยาดม 1. ยาดม & Salmon Sashimi 2. ยาดม & External Harddisk 3. ยาดม & Bluetooth Speaker

2. Bluetooth Speaker & Salmon Sashimi            -> lift= 1.006033, support = 0.630435, confidence = 0.743590	

3. Bluetooth Speaker & External Harddisk    -> lift= 1.023847, support = 0.608696, confidence = 0.823529

4. แก้วเก็บความเย็น & ยาดม -> lift= 1.027113, support = 0.608696, confidence = 0.736842

5. แก้วเก็บความเย็น & External Harddisk  

6. Salmon Sashimi & ขนมจีนน้ำยาปู

7. Salmon Sashimi & พลาสเตอร์บรรเทาปวดตราเสือ

8. External Harddisk & หม้อทอดไร้น้ํามัน

## Criteria
'lift' >= 2 &
'support' >= 0.1 &
'confidence' >= 0.5) 

In [ ]:
item_item_matrix = pd.DataFrame(index=basket_encode_final_df.columns,columns=basket_encode_final_df.columns)

In [ ]:
from scipy.spatial.distance import cosine

for i in range(0,len(item_item_matrix.columns)) :
    # Loop through the columns for each column
    for j in range(0,len(item_item_matrix.columns)) :
      # Fill in placeholder with cosine similarities
      item_item_matrix.iloc[i,j] = 1 -cosine(basket_encode_final_df.iloc[:,i],basket_encode_final_df.iloc[:,j])

In [ ]:
links = item_item_matrix.rename_axis('related item', 
                                     axis='columns').stack().reset_index()
links.columns = ['item', 'related item','value']

In [ ]:
links_filtered=links.loc[ (links['value'] > 0.1) & 
                         (links['item'] != links['related item']) ]

In [ ]:
import matplotlib.pyplot as plt
import networkx as nx

fig, ax=plt.subplots(figsize=(10,4))
GA=nx.from_pandas_edgelist(links_filtered,source='item',target='related item')
nx.draw(GA,with_labels=True)
plt.show()